editor/components/inserter/index.js

```diff
@@ -54,34 +101,63 @@ class Inserter extends Component {
 		}
 
 		return (
-			<Dropdown
-				className="editor-inserter"
-				position={ position }
-				onToggle={ this.onToggle }
-				expandOnMobile
-				headerTitle={ title }
-				renderToggle={ ( { onToggle, isOpen } ) => (
-					<IconButton
-						icon="insert"
-						label={ __( 'Add block' ) }
-						onClick={ onToggle }
-						className="editor-inserter__toggle"
-						aria-haspopup="true"
-						aria-expanded={ isOpen }
-					>
-						{ children }
-					</IconButton>
-				) }
-				renderContent={ ( { onClose } ) => {
-					const onSelect = ( item ) => {
-						onInsertBlock( item );
-
-						onClose();
-					};
-
-					return <InserterMenu onSelect={ onSelect } />;
-				} }
-			/>
+			<Fragment>
+				{ this.state.mediaLibraryOpen &&
+					<MediaUpload
+						type="image"
+						onSelect={ ( { type, url, alt, width, height } ) => {
+							onInsertInline( { type, url, alt, width, height } );
+							this.setState( { mediaLibraryOpen: false } );
+						} }
+						onClose={ () => ( this.setState( { mediaLibraryOpen: false } ) ) }
+						render={ ( { open } ) => {
+							open();
+							return null;
+						} }
+					/>
+				}
+				<Dropdown
+					className="editor-inserter"
+					position={ position }
+					onToggle={ this.onToggle }
+					expandOnMobile
+					headerTitle={ title }
+					renderToggle={ ( { onToggle, isOpen } ) => (
+						<IconButton
+							icon="insert"
+							label={ __( 'Add block' ) }
+							onClick={ onToggle }
+							className="editor-inserter__toggle"
+							aria-haspopup="true"
+							aria-expanded={ isOpen }
+						>
+							{ children }
+						</IconButton>
+					) }
+					renderContent={ ( { onClose } ) => {
+						const onSelect = ( item ) => {
+							onInsertBlock( item );
+
+							onClose();
+						};
+						const onImageSelect = () => {
+							this.setState( { mediaLibraryOpen: true } );
+
+							onClose();
+						};
+
+						if ( this.state.isInline ) {
```

Comment:
This is all very image specific? Would it make sense to generalise? Maybe good for a first iteration though...

Reply:
That’s a good question, @jayshenk – do you think we should aim to merge this before or adding other inline blocks like mentions?

Reply:
Good question. I have been studying the autocompleters, of which currently there are just two: the user autocomplete (@ mention) and the block autocomplete (slash command). Since it has been decided not to show the Inline Blocks menu when a paragraph (the default block) is empty, I think this excludes the possibility of having the slash command as an inline block. Also the default block is the only one in which the slash command can be used. That leaves the @ mention, which I think I could include in this PR.

I have been studying the autocomplete to see how involved this would be. Need to look into it more but I think it should be doable. Perhaps there could be an inline blocks module that exports both the inline image and @ mention, but I need to do some further planning to decide the best way to structure this. Perhaps there would be a `type` property of inline blocks, which could be `image` (or something more generic?) or `autocomplete`, with the possibility of adding more types in the future.

The autocomplete type would need to trigger the autocomplete to run, which currently is triggered by `keypress` I believe, but now would need to be triggered when selected in the inserter. We would also need to think about if we would want to make this extensible, allowing plugins to add new inline blocks.

Reply:
https://github.com/WordPress/gutenberg/pull/5794/commits/06ce12556d7bfae783db4a857edca002ec1b91e5 adds an inline blocks module, in preparation for adding other inline blocks like @ mentions.